### PR TITLE
[Ftr] tracing

### PR DESCRIPTION
**Pull request by LvBay**
**What this PR does**:Support jaeger tracing

**Which issue(s) this PR fixes**:

Fixes #

**Special notes for your reviewer**:

**Does this PR introduce a user-facing change?**:

```release-note

```

### CODE REVIEW

File: pkg/client/dubbo/dubbo.go [hunk 1]
```diff
@@ -22,20 +22,20 @@ import (
...
-import (
+	"go.opentelemetry.io/otel/trace"
+
 	"github.com/apache/dubbo-go/common/constant"
 	dg "github.com/apache/dubbo-go/config"
 	"github.com/apache/dubbo-go/protocol/dubbo"
-	fc "github.com/dubbogo/dubbo-go-pixiu-filter/pkg/api/config"
-	"github.com/pkg/errors"
-)
+	"go.opentelemetry.io/otel"
```
Comment (ztelur) [inline pkg/client/dubbo/dubbo.go:31]: pls split this import block into three part:
- first block contains all Go build-in package;
- second block contains third party package;
- third block contains package in project

File: pkg/filter/tracing/tracing.go [hunk 2]
```diff
@@ -0,0 +1,120 @@
...
+		spanName := spanNamePrefix + hc.Request.Method + " " + hc.Request.URL.Path
+		tr := otel.Tracer(traceName)
+		ctx := extractTraceCtxRequest(hc.Request)
+		ctxWithTid, span := tr.Start(ctx, spanName)
+
+		body := contexthttp.ExtractRequestBody(hc.Request)
+		span.SetAttributes(attribute.Key(spanTagBody).String(string(body)))
+		hc.Request = hc.Request.WithContext(ctxWithTid)
+		hc.Next()
+		span.End()
```
Comment (ztelur) [inline pkg/filter/tracing/tracing.go:104]: immediately end span ?

File: pkg/context/http/util.go [hunk 3]
```diff
@@ -84,3 +88,31 @@ func HttpRouteActionMatch(c *HttpContext, ra model.RouteAction) bool {
 
 	return true
 }
+
+// 将request.body写入到span中，并重新放回去
```
Comment (tydhot) [inline pkg/context/http/util.go:92]: I think we should avoid Chinese

Comment (AlexStocks) [reply pkg/context/http/util.go:92]: agree with u

File: pkg/context/http/util.go [hunk 4]
```diff
@@ -84,3 +88,31 @@ func HttpRouteActionMatch(c *HttpContext, ra model.RouteAction) bool {
...
+	}
+	body, err := ioutil.ReadAll(req.Body)
+	if err != nil {
+		return nil
+	}
+	req.Body = ioutil.NopCloser(bytes.NewBuffer(body))
+	return body
+}
+
+// 是否为文件上传 upload方法
```
Comment (AlexStocks) [inline pkg/context/http/util.go:106]: using english comment, pls.

File: pkg/context/http/util.go [hunk 5]
```diff
@@ -84,3 +88,31 @@ func HttpRouteActionMatch(c *HttpContext, ra model.RouteAction) bool {
...
+func isUpload(req *http.Request) bool {
+	isUpload := false
```
Comment (AlexStocks) [inline pkg/context/http/util.go:108]: var isUpload bool 就可以了。其默认值就是 false

File: pkg/filter/tracing/tracing.go (hunk 2 above)
Comment (LvBay) [reply pkg/filter/tracing/tracing.go:104]: 先执行hc.Next()，再span.End()

File: pkg/client/dubbo/dubbo.go (hunk 1 above)
Comment (LvBay) [reply pkg/client/dubbo/dubbo.go:31]: okay

Comment (LvBay) [reply pkg/client/dubbo/dubbo.go:31]: One extra question, why not use goimports or goreturn of the go community?

File: pkg/filter/tracing/tracing.go (hunk 2 above)
Comment (ztelur) [reply pkg/filter/tracing/tracing.go:104]: oh, my fault

File: pkg/client/dubbo/dubbo.go (hunk 1 above)
Comment (ztelur) [reply pkg/client/dubbo/dubbo.go:33]: we can use tool like this https://github.com/apache/dubbo-go/pull/1346 later

File: pkg/filter/tracing/tracing.go [hunk 6]
```diff
@@ -0,0 +1,138 @@
...
+
+package tracing
+
+import (
+	"context"
+	"log"
+	"net/http"
+	"time"
+
+	"go.opentelemetry.io/otel/trace"
```
Comment (AlexStocks) [inline pkg/filter/tracing/tracing.go:26]: split it

File: pkg/filter/tracing/tracing.go [hunk 7]
```diff
@@ -0,0 +1,138 @@
...
+func (f tracerFilter) Do() fc.FilterFunc {
+	return func(c fc.Context) {
+		hc := c.(*contexthttp.HttpContext)
+		spanName := spanNamePrefix + hc.Request.Method + " " + hc.Request.URL.Path
+		tr := otel.Tracer(traceName)
+		ctx := extractTraceCtxRequest(hc.Request)
+		ctxWithTid, span := tr.Start(ctx, spanName)
+
+		body := contexthttp.ExtractRequestBody(hc.Request)
+		span.SetAttributes(attribute.Key(spanTagBody).String(string(body)))
```
Comment (tydhot) [inline pkg/filter/tracing/tracing.go:119]: 我个人觉得没必要去记录请求body 存储比较消耗资源了 也属于比较敏感的数据

Comment (cityiron) [reply pkg/filter/tracing/tracing.go:119]: I agree.

Comment (LvBay) [reply pkg/filter/tracing/tracing.go:119]: 那就不记录body了~

File: pkg/filter/tracing/tracing.go [hunk 8]
```diff
@@ -0,0 +1,147 @@
...
+	default:
+		panic("unsupported tracing")
+	}
+
+	return TraceFilter{}
+}
+
+// Do execute tracerFilter filter logic.
+func (f TraceFilter) Handle(hc *contexthttp.HttpContext) {
+	spanName := spanNamePrefix + hc.Request.Method + " " + hc.Request.URL.Path
```
Comment (tydhot) [inline pkg/filter/tracing/tracing.go:125]: 一般情况下在网关层spanName直接到method就可以了，不需要把url path作为spanName。这是一个会无限膨胀的字段，会给后期聚合带来非常大的困难。

File: pkg/context/http/util.go [hunk 9]
```diff
@@ -60,3 +64,31 @@ func HttpRouteActionMatch(c *HttpContext, ra model.RouteAction) bool {
 
 	return true
 }
+
+// ExtractRequestBody extract body of http request
+func ExtractRequestBody(req *http.Request) []byte {
```
Comment (tydhot) [inline pkg/context/http/util.go:69]: 我还是建议不要采集body了，如果这是个上传文件的请求，或是敏感请求，不是很合适

**Review by ztelur — APPROVED** [review]

**Review by AlexStocks — APPROVED** [review]

**Review by xiaoliu10 — APPROVED** [review]

**Pull request merged**